### PR TITLE
Extension library

### PR DESCRIPTION
**Pull request by mrambacher**
This is the first pass of the changes for dynamically adding extensions to the RocksDB at run-time.  These changes were discussed at the RocksDB meetup and are not ready for general release.  After the meetup, I attempted to make the code more closely follow the ObjectRegistry pattern, but think the existing code cannot be fully extended to meet our goals/requirements.

A "DynamicLibrary" class was added to the Env object that loads a shared library.  An ExtensionLoader class was added (off the DBOptions) that stores the loaded libraries.

An "Extension" base class was added.  In this PR, the Env, EventListener, and TableFactory classes are Extension derivations (MergeOperator, Comparator, CompactionFilter are other prime candidates to support Rocksandra).  An ExtensionLoader can create an instance of an extension by knowing its name and type.  An Extension can be configured via ConfigureFromString/Map, where individual Extension properties are set via the SetOption method.   Extension instances can also be validated via a SanitizeOptions method (these methods were based on TableFactory).  Tests were added to test that these objects could be created and configured via this mechanism (TableFactory tests are pending).

The end goal is to be able to load extensions from an INI file.  An INI file can specify the libraries that contain extensions and the names and configurations of these extensions.  

Outside of the review of the basic code/functionality, there are a couple "next steps":
-> Read and create extensions from an INI file
-> Be able to write extension configurations to an INI file
-> Further test/validate this pattern for Rocksandra by removing the Cassandra-specific code from
JNI/libraries and adding the appropriate new Java APIs.

One outstanding design issue is what (if any) level of security should be added to the library loading mechanism.  I also think there needs to be some validation that the dynamic extension library was built for the right RocksDB release.

Feedback appreciated.

### CODE REVIEW

File: db/db_options_test.cc [hunk 1]
```diff
@@ -514,7 +514,7 @@ TEST_F(DBOptionsTest, SetStatsDumpPeriodSec) {
   }
 }
 
-TEST_F(DBOptionsTest, RunStatsDumpPeriodSec) {
+TEST_F(DBOptionsTest, DISABLED_RunStatsDumpPeriodSec) {
```
Comment (riversand963) [inline db/db_options_test.cc]: Do you plan to enable these tests again?

File: foo [hunk 2]
```diff
@@ -0,0 +1,3 @@
+Makefile:137: Warning: Compiling in debug mode. Don't use the resulting binary in production
```
Comment (riversand963) [inline foo]: Remove this file?

File: extensions/#extension_loader.cc# [hunk 3]
```diff
@@ -0,0 +1,125 @@
+// Copyright (c) 2011-present, Facebook, Inc. All rights reserved.
```
Comment (riversand963) [inline extensions/#extension_loader.cc#]: Should this file be removed?

**Comment by mrambacher** [comment]
Closing  this PR, as it is also subsumed by newer, smaller ones.

**Pull request closed without merging**